strawberryfields/configuration.py

```diff
@@ -0,0 +1,329 @@
+# Copyright 2019-2020 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+r"""
+This module contains functions used to load, store, save, and modify
+configuration options for Strawberry Fields.
+"""
+import logging as log
+import os
+
+import toml
+from appdirs import user_config_dir
+
+log.getLogger()
+
+DEFAULT_CONFIG_SPEC = {
+    "api": {
+        "authentication_token": (str, ""),
+        "hostname": (str, "platform.strawberryfields.ai"),
+        "use_ssl": (bool, True),
+        "port": (int, 443),
+    }
+}
+
+
+class ConfigurationError(Exception):
+    """Exception used for configuration errors"""
+
+
+def load_config(filename="config.toml", **kwargs):
+    """Load configuration from keyword arguments, configuration file or
+    environment variables.
+
+    .. note::
+
+        The configuration dictionary will be created based on the following
+        (order defines the importance, going from most important to least
+        important):
+
+        1. keyword arguments passed to ``load_config``
+        2. data contained in environmental variables (if any)
+        3. data contained in a configuration file (if exists)
+
+    Kwargs:
+        filename (str): the name of the configuration file to look for
+
+        Additional configuration options are detailed in
+        :doc:`/introduction/configuration`
+
+    Returns:
+        dict[str, dict[str, Union[str, bool, int]]]: the configuration
+    """
+    config = create_config()
+
+    filepath = get_config_filepath(filename=filename)
+
+    if filepath is not None:
+        loaded_config = load_config_file(filepath)
+        valid_api_options = keep_valid_options(loaded_config["api"])
+        config["api"].update(valid_api_options)
+    else:
+        log.info("No Strawberry Fields configuration file found.")
+
+    update_from_environment_variables(config)
+
+    valid_kwargs_config = keep_valid_options(kwargs)
+    config["api"].update(valid_kwargs_config)
+
+    return config
+
+
+def create_config(authentication_token="", **kwargs):
```

Comment:
To make codefactor have one less issue:
```suggestion
def create_config(authentication_token=None, **kwargs):
```